### DISCUSSION TITLE
有server支持多线程 的例子吗

### DISCUSSION THREAD

**Discussion by magic000** [Tencent/tquic#110] [General]
你好，有多线程服务端的压测数据吗
想测试下 多线程情况下， server 持续发送数据1Mb/s  给客户端， 看看这种场景下的性能
背景是 评估直播场景下 这个库的性能怎么样

**Comment by wangfuyu** [comment]
> 你好，有多线程服务端的压测数据吗 想测试下 多线程情况下， server 持续发送数据1Mb/s 给客户端， 看看这种场景下的性能 背景是 评估直播场景下 这个库的性能怎么样

TQUIC可集成到多进程或多线程框架中。

直播数据流，很多时间是没有数据的（帧与帧之间有gap），也就是说拥塞控制模型常常会进入app-limited状态，可以等同于长连接上的大量的小文件请求（多路复用，但流之间有间隔）。
如果您这边说的『性能』是指单机性能而不是传输性能，那您可以直接参考：https://tquic.net/docs/further_readings/benchmark

**Comment by magic000** [comment]
是想压测传输性能，  quic-go 我是写一个server， 收到请求后 就模拟直播的流 定时给客户端发送假数据，单路流带宽保持在1Mb/s
想对比下rust quic，但是测试的例子 只能利用单核并且是发送文件， 需要改造成多线程 并且  改成构造假数据 按一定的速率发送
不过我还不了解rust的语法，像看看你们还有其他例子 可以参考的吗

**Comment by wangfuyu** [comment]
> 是想压测传输性能， quic-go 我是写一个server， 收到请求后 就模拟直播的流 定时给客户端发送假数据，单路流带宽保持在1Mb/s 想对比下rust quic，但是测试的例子 只能利用单核并且是发送文件， 需要改造成多线程 并且 改成构造假数据 按一定的速率发送 不过我还不了解rust的语法，像看看你们还有其他例子 可以参考的吗

根据我个人的理解，从传输角度看，直播一般关注的不是传输性能（吞吐），而是传输效果，包括但不限于：成功率、首帧耗时（或者秒开率）、卡顿率（含次数和时长）、延迟等； 毕竟直播是要实时的数据，并不是说下载速率越大，传输效果就越好，对吧？
另外，若是关注传输性能，非单机性能，那和是否『多线程』没有太大关系呢？多线程或者多进程只是提升单机并发处理的能力。

传输性能/效果，和拥塞控制算法及相关参数的设置关系比较大，和实现的语言以及是否多线程关系都不大哈。同样的拥塞控制算法，比如BBR，只要算法实现准确，传输效果得靠各使用者根据自己的业务模型对算法进行相应的调整来达成，不是各开源QUIC库的默认配置（不同开源库中的相同拥塞控制算法初始参数不一样，实际应用时，都需要使用者根据自己需要来更新配置）就能适配所有场景哈。

最后： 我们后续会将部分业务场景，用各开源库的各个常见拥塞算法（设置相同的参数）测试效果（传输性能），放出来供大家参考，您这边可以关注下哈。

**Comment by magic000** [comment]
我们比较关注机器成本， quic消耗CPU 本身就比较多， 想对比下不同的语言 在io发送性能的差距大不大

**Comment by wangfuyu** [comment]
> 我们比较关注机器成本， quic消耗CPU 本身就比较多， 想对比下不同的语言 在io发送性能的差距大不大

根据您上面提到的信息，实际上您关注的是单机性能，而非传输性能（吞吐），那您可以直接参考我们对业界几个比较优秀的开源库的性能测试数据哈，链接：https://tquic.net/zh/docs/further_readings/benchmark/

quic-go这个库我们没有在上面链接中做对比，主要是这个早几年大家都已经有比较明确的结论了，您可以在网上搜索到很多参考数据哈。

另外提下，TQUIC是一个实现QUIC协议栈的库，socket相关读写是由集成的框架负责的哈。

**Comment by magic000** [comment]
好的，我再看看，感谢
